bliss/datasets/catsim.py

```diff
@@ -82,151 +41,186 @@ def __init__(
+        image_shape = (self.n_bands, self.slen, self.slen)
+        self.background = np.zeros(image_shape, dtype=self.dtype)
+        for b in range(self.n_bands):
+            self.background[b, :, :] = background[b]
+
+        # Get unintegrated galsim psf for the convolution
+        psf = fits.getdata(psf_file)
+        assert len(psf.shape) == 2, "Same PSF for all bands for now."
+        psf = psf - psf[1, int(psf.shape[1] / 2)] * 2
+        psf[psf < 0] = 0
+        psf /= np.sum(psf)
+        self.psf = galsim.InterpolatedImage(
+            galsim.Image(psf), scale=self.pixel_scale
+        ).withFlux(1.0)
 
     def center_deviation(self, entry):
         # random deviation from exactly in center of center pixel, in arcseconds.
         deviation_ra = (np.random.rand() - 0.5) if self.deviate_center else 0.0
         deviation_dec = (np.random.rand() - 0.5) if self.deviate_center else 0.0
-        entry["ra"] = deviation_ra * self.pixel_scale  # arcseconds
+        entry["ra"] = deviation_ra * self.pixel_scale
         entry["dec"] = deviation_dec * self.pixel_scale
         return entry
 
     def render(self, entry):
         """
         Return a multi-band image corresponding to the entry from the catalog given.
 
-        * The final image includes its background based on survey's sky level.
+        * The final image includes background.
         * If deviate_center==True, then galaxy not aligned between bands.
         """
         image = np.zeros((self.n_bands, self.slen, self.slen), dtype=self.dtype)
+        entry = self.center_deviation(entry)  # all bands are centered same way.
 
-        with SurveyObs(self) as obs:
-            for i, band in enumerate(self.bands):
-                entry = self.center_deviation(entry)
-                image_no_background = self.single_band(entry, obs[i], band)
-                image[i, :, :] = image_no_background + self.background[i]
+        for b, band in enumerate(self.bands):
+            image[b] = self.single_band_galaxy(entry, band).array
 
+        # add background and (optionally) Gaussian noise
+        image += self.background
+        _image = np.sqrt(image) * np.random.randn(*image.shape) * self.noise_factor
+        image += _image
         return image, self.background
 
-    def single_band(self, entry, single_obs, band):
-        """Builds galaxy from a single entry in the catalog. With no background sky level added."""
-        import descwl
-        import galsim
+    def get_flux(self, ab_magnitude):
+        """Convert source magnitude to flux.
+        Args:
+            ab_magnitude(float): AB magnitude of source.
+        Returns:
+            float: Flux in detected electrons.
+        """
 
-        galaxy_builder = descwl.model.GalaxyBuilder(
-            single_obs, no_disk=False, no_bulge=False, no_agn=False, verbose_model=False
-        )
+        # TODO: other bands that are not 'r'
+        return 3000 * 10 ** (-0.4 * (ab_magnitude - 24))
```

Comment:
right now I just played with the numbers to get something like the fluxes in SDSS, is there a more principled way ?

Reply:
That seems fine. The results shouldn’t be too sensitive to differences in the distribution of fluxes between the simulated data and the real data. Probably better to error on the side of making the simulated too diverse